bot/cogs/moderation/incidents.py

```diff
@@ -0,0 +1,334 @@
+
+        We do not release `event_lock` until we receive the corresponding `message_delete` event.
+        This ensures that if there is a racing event awaiting the lock, it will fail to find the
+        message, and will abort. There is a `timeout` to ensure that this doesn't hold the lock
+        forever should something go wrong.
+        """
+        members_roles: t.Set[int] = {role.id for role in member.roles}
+        if not members_roles & ALLOWED_ROLES:  # Intersection is truthy on at least 1 common element
+            log.debug(f"Removing invalid reaction: user {member} is not permitted to send signals")
+            await incident.remove_reaction(reaction, member)
+            return
+
+        if reaction not in ALLOWED_EMOJI:
+            log.debug(f"Removing invalid reaction: emoji {reaction} is not a valid signal")
+            await incident.remove_reaction(reaction, member)
+            return
+
+        # If we reach this point, we know that `emoji` is a `Signal` member
+        signal = Signal(reaction)
+        log.trace(f"Received signal: {signal}")
+
+        if signal not in (Signal.ACTIONED, Signal.NOT_ACTIONED):
+            log.debug("Reaction was valid, but no action is currently defined for it")
+            return
+
+        relay_successful = await self.archive(incident, signal)
+        if not relay_successful:
+            log.trace("Original message will not be deleted as we failed to relay it to the archive")
+            return
+
+        timeout = 5  # Seconds
+        confirmation_task = self.make_confirmation_task(incident, timeout)
+
+        log.trace("Deleting original message")
+        await incident.delete()
+
+        log.trace(f"Awaiting deletion confirmation: {timeout=} seconds")
+        try:
+            await confirmation_task
+        except asyncio.TimeoutError:
+            log.warning(f"Did not receive incident deletion confirmation within {timeout} seconds!")
+        else:
+            log.trace("Deletion was confirmed")
+
+    async def resolve_message(self, message_id: int) -> t.Optional[discord.Message]:
+        """
+        Get `discord.Message` for `message_id` from cache, or API.
+
+        We first look into the local cache to see if the message is present.
+
+        If not, we try to fetch the message from the API. This is necessary for messages
+        which were sent before the bot's current session.
+
+        In an edge-case, it is also possible that the message was already deleted, and
+        the API will respond with a 404. In such a case, None will be returned.
+        This signals that the event for `message_id` should be ignored.
+        """
+        await self.bot.wait_until_guild_available()  # First make sure that the cache is ready
+        log.trace(f"Resolving message for: {message_id=}")
+        message: discord.Message = self.bot._connection._get_message(message_id)
+
+        if message is not None:
```

Comment:
```suggestion
        if message:
```

Reply:
Also, type hint should `Optional[discord.Message]` ad L268

Reply:
I generally prefer checking for `None` explicitly. I want to return if `message` is not `None`, not if `message` is truthy. I don't feel like "Message can never be falsey, so we can get away with this" is a good enough justification - explicit is better than implicit, after all - but yeah, I suppose it's more of a personal preference. If the check was implicit, and I were reviewing this and didn't know whether a `Message` can ever be falsey, I'd have to go searching for it. This way, I don't.

Regarding the annotation, good catch, that is incorrect. I'll adjust it.

Reply:
Annotation fixed in 8a08ca3.